### DISCUSSION TITLE
Is it Server OR WASM or can both be used in same app?

### DISCUSSION THREAD

**Discussion by MPatriot** [oqtane/oqtane.framework#3243] [Q&A]
Probably another basic / obvious question but can both Server mode and WASM mode / enviroment be used in same app?

For SEO purposes and general browsing it seems Server mode is best both for speed and searchability.  On the other hand for more CPU / Memory intensive things that SEO isn't required or desired for or even admin stuff I'd rather use WASM.

Just wondering if / how both can be used / implemented?

Thanks

**Comment by sbwalker** [comment]
@MPatriot Oqtane supports both Server and Wasm hosting models in the same "installation". Installations are multi-tenant - which means that multiple sites can be hosted from the same install. Each site can specify its own hosting model. But you currently cannot mix hosting models within the same site.

**Reply by thabaum** [reply]
.NET 8 should allow mixing Server/WebAssembly "components" that can load both in one page regardless of server or wasm if I am not mistaken.  You declare which or "Auto" mode to allow the framework to decide.  

[https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.components.web.rendermode?view=aspnetcore-8.0](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.components.web.rendermode?view=aspnetcore-8.0)

From my understanding you will start in "Server Mode" then it silently loads things in background while browsing the site and uploads the "WASM" files. 

Next version of Oqtane should cure any issues relating to how you may want things to load.  Not 100% sure this is what you are after but worth getting ready for :)

[Auto](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.components.web.rendermode.auto?view=aspnetcore-8.0#microsoft-aspnetcore-components-web-rendermode-auto) 	

Gets an [IComponentRenderMode](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.components.icomponentrendermode?view=aspnetcore-8.0) that means the render mode will be determined automatically based on a policy.

[Server](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.components.web.rendermode.server?view=aspnetcore-8.0#microsoft-aspnetcore-components-web-rendermode-server) 	

Gets an [IComponentRenderMode](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.components.icomponentrendermode?view=aspnetcore-8.0) that represents rendering interactively on the server via Blazor Server hosting with server-side prerendering.

[WebAssembly](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.components.web.rendermode.webassembly?view=aspnetcore-8.0#microsoft-aspnetcore-components-web-rendermode-webassembly) 	

Gets an [IComponentRenderMode](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.components.icomponentrendermode?view=aspnetcore-8.0) that represents rendering interactively on the client via Blazor WebAssembly hosting with server-side prerendering.

Pretty cool stuff to look forward to ;)

**Reply by sbwalker** [reply]
@thabaum please note that the .NET 8 features will not automatically work in Oqtane. I have been discussing this topic with Microsoft since February because the approach they took is dependent upon static server-side routing and the default Blazor router - however Oqtane uses dynamic routing and has its own custom router to support multi-tenancy and modularity. At this point it is not clear if Oqtane will be able to leverage these new features in .NET 8.

**Reply by thabaum** [reply]
Sounds like it may be a challenge to make it work in the custom router with out a lot of testing of all the scenarios within Oqtane.  Hopefully some scenarios if not all can be supported at some point once potential problem areas within the custom router can be identified.

For most projects you just need to update the runtime in each project and your off and running with the next version.  Is this the case with Oqtane?  I would not mind putting a version together running Oqtane 5 preview so to speak.

**Reply by sbwalker** [reply]
Migrating Oqtane to .NET 8 will not be a challenge, as .NET 8 still supports the current Server, WebAssembly, and Hybrid hosting models. So all of the current Oqtane functionality will work on .NET 8. However the new Server-Side Rendering (SSR) mode (which is sometimes called "Blazor United") requires a different approach to building your application. So it remains to be seen if there is any way for Oqtane to bridge the gap. For example,  Blazor United relies on static server-side routes - which means that it does not support the ability for you to add routes, modify route paths, or delete routes at run-time (which is fundamental to the Oqtane site builder experience - and why Oqtane uses a custom router). So if you wanted to add a page to your site, you would need to restart your application so that the new static route could be added to the route table at startup.

**Reply by thabaum** [reply]
Does this render mode need to be included in pages?  I was thinking it would be a component that handles the new render modes within a page.  So regardless if you use a dynamically created page or not it would push things to the browser as desired for each component independently set by the mode desired for it.

Seems almost like docker images having to all be updated with same code, you would need to upload the new version to the client in a wasm package.

Also I was thinking this just pushed the .NET runtime to the client loading as Server-Side first then running Client-Side as the dll package now resides with the client to run .NET 8 from watching a few conferences about it.  So you dont have to keep downloading it each page click, just the first one.  So maybe it can be partially supported in Oqtane if not fully.

**Reply by sbwalker** [reply]
The render mode can be set by individual components but they are still dependent on static server-side routing and the Blazor default router. Your understanding of how it works from a technical perspective is not accurate. Basically on startup the server interrogates the razor components server-side to find their routes, builds a static route table, pre-renders the components server-side, passes the route table to the Blazor router, and when the Blazor router processes a page route which exists in the route table, it short circuits the interactive component rendering process (ie. and relies on the pre-rendered server-side output).

**Reply by thabaum** [reply]
[https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-preview-7/#auto-render-mode](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-preview-7/#auto-render-mode)

I have been trying to follow what @danroth27 says in webinars and blogs to see how things are developing with .NET 8.

> The new Auto interactive render mode for Blazor web apps combines the strengths of the Server and WebAssembly render modes into a single dynamic option. The Auto render mode uses WebAssembly-based rendering if the .NET WebAssembly runtime can be loaded quickly (within 100ms). This typically is the case when the runtime was previously downloaded and cached, or when using a high speed network. Otherwise, the Auto render mode falls back to using the Server render mode while the .NET WebAssembly runtime is downloaded in the background.

I guess this is what I am talking about from what I heard him talk about in a previous webinar which I think its the one he calls Oqtane a CMS :)

Also this is the other part from that page that is of interest here:

Improved integration of interactive components with server-side rendering

> Blazor in .NET 8 has advanced server-side rendering capabilities, like enhanced navigation and form handling. This preview improves the integration of interactive components with server-side rendering. Enhanced navigation, enhanced form handling, and streaming rendering can now add and remove interactive components and set parameters on them.

The "streaming rendering" feature is what we may have issues with in Oqtane if I am understanding this correctly from your viewpoint?

This is a little off topic, but the new set of identity API endpoints may allow these components to be moved to the client project using a service to talk to these new server API endpoints?

And one last interesting thing is the new anti-forgery middle-ware that can be included after Authentication and Authorization, but off topic not starting a discussion just mentioning.

Back to the topic here, CPU loads and so on I swear I remember hearing something about having access to utilizing multiple CPU threads.

**Comment by sbwalker** [comment]
@thabaum both the Auto mode and streaming rendering scenarios rely on static routes so they are both challenges for Oqtane. I spoke to the Microsoft PM about the new Identity endpoints and for the most part they are simply wrappers over top of the current Identity API. The major difference is that the Identity implementation in Oqtane has been enhanced to be multi-tenant (whereas the default Identity API does not support multi-tenancy). So it does not make sense for Oqtane to use the new Identity endpoints. Also Oqtane has included native Razor components as a Blazor UI for the Identity Login, User, Roles, etc... capabilities for the past 3 years  (whereas the default Identity used server-side Razor pages). So it does not make sense for Oqtane to use the new Identity UIs either. And Oqtane adopted a BFF security architecture 2 years ago when support for OIDC was added, so it is good to see that Microsoft is also now adopting a BFF architecture for Identity in .NET 8. On the topic of ant-forgery, Oqtane already uses anti-forgery by default on all POST, PUT, DELETE requests.

**Reply by thabaum** [reply]
@sbwalker Sounds like the Oqtane Framework is ahead of the game in a few areas such as those explained so things should just keep trucking along. Thanks for all the deep thought put into all these aspects of the framework.  I know all of this has been discussed in the past, just with new features coming to .NET 8 and release in about 2 months it is hard not to get a little excited thinking things may have changed from earlier previews to the latest to try to think of how any of these can potentially help advance the Oqtane Framework.  The complexity of the Oqtane Framework goes beyond any tutorial or course I have ever taken or can find on Microsoft's site ;).  Thanks again for taking the time to explain in further detail.
